src/compiler/moduleSpecifiers.ts

```diff
@@ -5,6 +5,47 @@ namespace ts.moduleSpecifiers {
         readonly importModuleSpecifierPreference?: "relative" | "non-relative";
     }
 
+    const enum RelativePreference { Relative, NonRelative, Auto }
+    // Determines whether we import `foo/index.ts` as "foo", "foo/index", or "foo/index.js"
+    const enum Ending { Minimal, Index, JsExtension }
+
+    // Processed preferences
+    interface Preferences {
+        readonly relativePreference: RelativePreference;
+        readonly ending: Ending;
+    }
+
+    function getPreferences({ importModuleSpecifierPreference }: ModuleSpecifierPreferences, compilerOptions: CompilerOptions, importingSourceFile: SourceFile): Preferences {
+        return {
+            relativePreference: importModuleSpecifierPreference === "relative" ? RelativePreference.Relative : importModuleSpecifierPreference === "non-relative" ? RelativePreference.NonRelative : RelativePreference.Auto,
+            ending: usesJsExtensionOnImports(importingSourceFile) ? Ending.JsExtension
+                : getEmitModuleResolutionKind(compilerOptions) !== ModuleResolutionKind.NodeJs ? Ending.Index : Ending.Minimal,
+        };
+    }
+
+    function getPreferencesForUpdate(_: ModuleSpecifierPreferences, compilerOptions: CompilerOptions, oldImportSpecifier: string): Preferences {
+        return {
+            relativePreference: isExternalModuleNameRelative(oldImportSpecifier) ? RelativePreference.Relative : RelativePreference.NonRelative,
+            ending: fileExtensionIs(oldImportSpecifier, Extension.Js) ? Ending.JsExtension

```

Comment:
Need to handle js and jsx both since that's what tryResolve(Extensions.Javascript) does.